controller/integration/test/controlflow/controlflow_test.go

```diff
@@ -0,0 +1,232 @@
+// +build k8srequired
+
+package controlflow
+
+import (
+	"reflect"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cenkalti/backoff"
+	"github.com/giantswarm/apiextensions/pkg/apis/core/v1alpha1"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/operatorkit/controller"
+	"github.com/giantswarm/operatorkit/controller/integration/testresource"
+	"github.com/giantswarm/operatorkit/controller/integration/wrapper"
+	"github.com/giantswarm/operatorkit/controller/integration/wrapper/nodeconfig"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+)
+
+// Test_Finalizer_Integration_Controlflow is a integration test to
+// check that errors during deletion prevent the finalizer from removal.
+func Test_Finalizer_Integration_Controlflow(t *testing.T) {
+	objName := "test-obj"
+	testFinalizer := "operatorkit.giantswarm.io/test-operator"
+	testNamespace := "finalizer-integration-reconciliation-test"
+	operatorName := "test-operator"
+
+	var err error
+	var tr *testresource.Resource
+	{
+		c := testresource.Config{}
+
+		tr, err = testresource.New(c)
+		if err != nil {
+			t.Fatal("expected", nil, "got", err)
+		}
+	}
+
+	resources := []controller.Resource{
+		controller.Resource(tr),
+	}
+
+	c := nodeconfig.Config{
+		Resources: resources,
+
+		Name:      operatorName,
+		Namespace: testNamespace,
+	}
+
+	nodeconfigWrapper, err := nodeconfig.New(c)
+	if err != nil {
+		t.Fatal("expected", nil, "got", err)
+	}
+
+	testWrapper := wrapper.Interface(nodeconfigWrapper)
+
+	testWrapper.MustSetup(testNamespace)
+	defer testWrapper.MustTeardown(testNamespace)
+
+	controller := testWrapper.Controller()
+
+	// We start the controller.
+	go controller.Boot()
```

Comment:
The real action starts here, the testcase is pretty straight forwards but a lot of checks in there to make sure everything is alright.

Steps taken here:
1. Boot framework
2. Create CR and check if the finalizer gets added
3. Set `ThrowError` to true
4. Delete CR and check that it's not removed and finalizer still exists
5. Set `ThrowError` to false
6. Check if the CR is completely gone

Reply:
Pleas no error throwing in Go, returning only! :P